File: VB/DashboardExporterApp/ExportControlProviders/FunnelItemExportControlProvider.vb
Imports System.Collections.Generic
Imports System.Linq
Imports DevExpress.DashboardCommon
Imports DevExpress.XtraCharts
Imports DevExpress.XtraReports.UI

Namespace DashboardExporterApp
	Public Class FunnelItemExportControlProvider
		Implements ICustomExportControlProvider

		Private ReadOnly dashboardItem As CustomDashboardItem

		Public Sub New(ByVal dashboardItem As CustomDashboardItem)
			Me.dashboardItem = dashboardItem
		End Sub
		Private Function ICustomExportControlProvider_GetPrintableControl(ByVal customItemData As CustomItemData, ByVal exportInfo As CustomItemExportInfo) As XRControl Implements ICustomExportControlProvider.GetPrintableControl
			Dim chart As New XRChart()
			chart.Legend.Visibility = DevExpress.Utils.DefaultBoolean.True
			AddHandler chart.CustomDrawSeriesPoint, AddressOf CustomDrawSeriesPoint
			Dim flatData As DashboardFlatDataSource = customItemData.GetFlatData(New DashboardFlatDataSourceOptions() With {.AddColoringColumns = True})
			ConfigureSeries(chart, customItemData, flatData, exportInfo.DrillDownValues.Count)
			SetSelectionMode(chart)
			SetSelection(chart, exportInfo.Selection, flatData)
			Return chart
		End Function
		Private Sub ConfigureSeries(ByVal chart As XRChart, ByVal customItemData As CustomItemData, ByVal flatData As DashboardFlatDataSource, ByVal drillDownLevel As Integer)
			chart.Series.Clear()
			Dim series As New Series("A Funnel Series", ViewType.Funnel)
			Dim values As IList(Of CustomItemBindingValue) = customItemData.GetBindings("Value")
			Dim arguments As IList(Of CustomItemBindingValue) = customItemData.GetBindings("Arguments")
			If values.Count > 0 AndAlso arguments.Count > 0 Then
				series.DataSource = flatData
				series.ValueDataMembers.AddRange(values(0).UniqueId)
				If dashboardItem.InteractivityOptions.IsDrillDownEnabled Then
					series.ArgumentDataMember = arguments(drillDownLevel).UniqueId
				Else
					series.ArgumentDataMember = arguments.Last().UniqueId
				End If
				series.ColorDataMember = flatData.GetColoringColumn(values(0).UniqueId).Name
			End If
			CType(series.Label, FunnelSeriesLabel).Position = FunnelSeriesLabelPosition.Center
			chart.Series.Add(series)
		End Sub
		Private Sub SetSelectionMode(ByVal chart As XRChart)
			chart.Chart.SeriesSelectionMode = SeriesSelectionMode.Point
			Select Case dashboardItem.InteractivityOptions.MasterFilterMode
				Case DashboardItemMasterFilterMode.None
					chart.Chart.SelectionMode = ElementSelectionMode.None
				Case DashboardItemMasterFilterMode.Single
					chart.Chart.SelectionMode = ElementSelectionMode.Single
				Case DashboardItemMasterFilterMode.Multiple
					chart.Chart.SelectionMode = ElementSelectionMode.Extended
				Case Else
					chart.Chart.SelectionMode = ElementSelectionMode.None
			End Select
		End Sub
		Private Sub SetSelection(ByVal chart As XRChart, ByVal selection As CustomItemSelection, ByVal flatData As DashboardFlatDataSource)
			For Each row As DashboardFlatDataSourceRow In selection.GetDashboardFlatDataSourceRows(flatData)
				chart.Chart.SelectedItems.Add(row)
			Next row
		End Sub
		Private Sub CustomDrawSeriesPoint(ByVal sender As Object, ByVal e As CustomDrawSeriesPointEventArgs)
			e.LabelText = e.SeriesPoint.Argument & " - " & e.LabelText
			e.LegendText = e.SeriesPoint.Argument
		End Sub
	End Class
End Namespace

File: CS/DashboardExporterApp/ExportControlProviders/FunnelItemExportControlProvider.cs
using System.Collections.Generic;
using System.Linq;
using DevExpress.DashboardCommon;
using DevExpress.XtraCharts;
using DevExpress.XtraReports.UI;

namespace DashboardExporterApp {
    public class FunnelItemExportControlProvider : ICustomExportControlProvider {
        readonly CustomDashboardItem dashboardItem;

        public FunnelItemExportControlProvider(CustomDashboardItem dashboardItem) {
            this.dashboardItem = dashboardItem;
        }
        XRControl ICustomExportControlProvider.GetPrintableControl(CustomItemData customItemData, 
            CustomItemExportInfo exportInfo) {
            XRChart chart = new XRChart();
            chart.Legend.Visibility = DevExpress.Utils.DefaultBoolean.True;
            chart.CustomDrawSeriesPoint += CustomDrawSeriesPoint;
            DashboardFlatDataSource flatData = customItemData.GetFlatData(new DashboardFlatDataSourceOptions() {
                AddColoringColumns = true });
            ConfigureSeries(chart, customItemData, flatData, exportInfo.DrillDownValues.Count);
            SetSelectionMode(chart);
            SetSelection(chart, exportInfo.Selection, flatData);
            return chart;
        }
        void ConfigureSeries(XRChart chart, CustomItemData customItemData,
            DashboardFlatDataSource flatData, int drillDownLevel) {
            chart.Series.Clear();
            Series series = new Series("A Funnel Series", ViewType.Funnel);
            IList<CustomItemBindingValue> values = customItemData.GetBindings("Value");
            IList<CustomItemBindingValue> arguments = customItemData.GetBindings("Arguments");
            if(values.Count > 0 && arguments.Count > 0) {
                series.DataSource = flatData;
                series.ValueDataMembers.AddRange(values[0].UniqueId);
                if(dashboardItem.InteractivityOptions.IsDrillDownEnabled)
                    series.ArgumentDataMember = arguments[drillDownLevel].UniqueId;
                else
                    series.ArgumentDataMember = arguments.Last().UniqueId;
                series.ColorDataMember = flatData.GetColoringColumn(values[0].UniqueId).Name;
            }
            ((FunnelSeriesLabel)series.Label).Position = FunnelSeriesLabelPosition.Center;
            chart.Series.Add(series);
        }
        void SetSelectionMode(XRChart chart) {
            chart.Chart.SeriesSelectionMode = SeriesSelectionMode.Point;
            switch(dashboardItem.InteractivityOptions.MasterFilterMode) {
                case DashboardItemMasterFilterMode.None:
                    chart.Chart.SelectionMode = ElementSelectionMode.None;
                    break;
                case DashboardItemMasterFilterMode.Single:
                    chart.Chart.SelectionMode = ElementSelectionMode.Single;
                    break;
                case DashboardItemMasterFilterMode.Multiple:
                    chart.Chart.SelectionMode = ElementSelectionMode.Extended;
                    break;
                default:
                    chart.Chart.SelectionMode = ElementSelectionMode.None;
                    break;
            }
        }
        void SetSelection(XRChart chart, CustomItemSelection selection, 
            DashboardFlatDataSource flatData) {
            foreach(DashboardFlatDataSourceRow row in selection.GetDashboardFlatDataSourceRows(flatData))
                chart.Chart.SelectedItems.Add(row);
        }
        void CustomDrawSeriesPoint(object sender, CustomDrawSeriesPointEventArgs e) {
            e.LabelText = e.SeriesPoint.Argument + " - " + e.LabelText;
            e.LegendText = e.SeriesPoint.Argument;
        }
    }
}
